### ISSUE TITLE
TinyGo support

### ISSUE THREAD

**Issue by bradleypeabody** [tinylib/msgp#294]
The msgp package currently does not compile under [TinyGo](https://tinygo.org).   Messagepack support on microcontrollers and in WASM I think is super useful and it would be great if msgp+tinygo worked.

The required fixes seem quite minimal - so far from my initial testing, just build tags:

```
diff --git a/msgp/advise_linux.go b/msgp/advise_linux.go
index 6c6bb37..7b6049f 100644
--- a/msgp/advise_linux.go
+++ b/msgp/advise_linux.go
@@ -1,4 +1,4 @@
-// +build linux,!appengine
+// +build linux,!appengine,!tinygo

 package msgp

diff --git a/msgp/file.go b/msgp/file.go
index 8e7370e..54586ba 100644
--- a/msgp/file.go
+++ b/msgp/file.go
@@ -1,5 +1,5 @@
 // +build linux darwin dragonfly freebsd netbsd openbsd
-// +build !appengine
+// +build !appengine,!tinygo

 package msgp

diff --git a/msgp/file_port.go b/msgp/file_port.go
index 6e654db..3e33b1f 100644
--- a/msgp/file_port.go
+++ b/msgp/file_port.go
@@ -1,4 +1,4 @@
-// +build windows appengine
+// +build windows appengine tinygo

 package msgp
```

And a couple for the `fwd` package as well:
```
diff --git a/writer_appengine.go b/writer_appengine.go
index e367f39..fa6e90a 100644
--- a/writer_appengine.go
+++ b/writer_appengine.go
@@ -1,4 +1,4 @@
-// +build appengine
+// +build appengine tinygo

 package fwd

diff --git a/writer_unsafe.go b/writer_unsafe.go
index a0bf453..1701572 100644
--- a/writer_unsafe.go
+++ b/writer_unsafe.go
@@ -1,4 +1,4 @@
-// +build !appengine
+// +build !appengine,!tinygo

 package fwd
```

If I do PRs to fix these things, is this something you would be able to merge?  I realize it may not be possible to get 100% of the functionality under TinyGo, but it seems like the vast majority of this can be made to work easily with just some build tags.  If you're good with the idea, I can do the work and send PRs.

**Comment by philhofer** [comment]
Yes, those changes seem perfectly fine. I'd be happy merging them.
(But even happier if we had some way to make sure we were covering these builds in our continuous integration so that we don't regress on this support once we have it.)

**Comment by bradleypeabody** [comment]
Awesome.

On the CI aspect, my thought was to do the following:

* `.travis.yml` would get updated to install `tinygo`
* `Makefile` would run something like `go test ./tinygotests`
* `tinygotests` would house regular Go tests that drive the execution of the `tinygo build` command against a set of subfolders.
* Individual tests could live in folders like `tinygotests/simpleparse` - which would have a complete program that gets compiled with tinygo
* For targets like `wasm` and various microcontrollers (e.g. `arduino-nano33`), it is simplest and most realistic to just test that the code compiles
* But `tinygo build` will also produce working Linux ELF executables, so test programs could be built with tinygo and then run and produce some expected output, which could be checked by the corresponding Go test that drives it from `tinygotests`.

I wish the setup were simpler, but I think if the complexity can be contained in one folder and just a couple of additional Makefile and travis lines like the above, it would be a workable way to go.

I can try this out and see how it goes. But I'll probably send you a simple PR first with just the low hanging fruit of the build tag changes.

**Comment by philhofer** [comment]
> I can try this out and see how it goes. But I'll probably send you a simple PR first with just the low hanging fruit of the build tag changes.

Sure, sounds good.

As far as CI goes: I'm fine with just testing that the code compiles. We already have `GOARCH=386` testing to stay 32-bit clean, at least.

**Comment by bradleypeabody** [comment]
Done, feature merged with #295
